state/sqlserver/migration.go

```diff
@@ -272,35 +271,85 @@ func (m *migration) createStoredProcedureIfNotExists(db *sql.DB, name string, es
 /* #nosec */
 func (m *migration) ensureUpsertStoredProcedureExists(db *sql.DB, mr migrationResult) error {
 	tsql := fmt.Sprintf(`
-		CREATE PROCEDURE %s (
-			@Key 			%s,
-			@Data 			NVARCHAR(MAX),
-			@RowVersion 	BINARY(8))
-		AS
-			IF (@RowVersion IS NOT NULL)
-			BEGIN
-				UPDATE [%s]
-				SET [Data]=@Data, UpdateDate=GETDATE()
-				WHERE [Key]=@Key AND RowVersion = @RowVersion
-
-				RETURN
-			END
-
-			BEGIN TRY
-				INSERT INTO [%s] ([Key], [Data]) VALUES (@Key, @Data);
-			END TRY
-
-			BEGIN CATCH
-				IF ERROR_NUMBER() IN (2601, 2627)
-				UPDATE [%s]
-				SET [Data]=@Data, UpdateDate=GETDATE()
-				WHERE [Key]=@Key AND RowVersion = ISNULL(@RowVersion, RowVersion)
-			END CATCH`,
+			CREATE PROCEDURE %s (
+				@Key 			%s,
+				@Data 			NVARCHAR(MAX),
+				@RowVersion		BINARY(8),
+				@FirstWrite		BIT)
+			AS
+				IF (@FirstWrite=1)
+					BEGIN
+						IF (@RowVersion IS NOT NULL)
+							BEGIN
+								BEGIN TRANSACTION;
+								IF NOT EXISTS (SELECT * FROM [%s] WHERE [KEY]=@KEY AND RowVersion = @RowVersion)
+									BEGIN
+										THROW 2601, ''FIRST-WRITE: COMPETING RECORD ALREADY WRITTEN.'', 1
+									END
```

Comment:
This is not always the case though: this case can also be because the row was never written before and the key does not exist.

Reply:
That's an excellent point. What do you suggest? Changing the error message?

If the record does not exist it's not actually possible to specify an ETAG (Row Version) to be written -- the stored procedure / DB won't allow the use of ETAGs other than to update an existing record.

Reply:
Yeah, I would just change the error message to something more general